### PR TITLE
add support for ViewHolder inside EfficientAdapter

### PR DESCRIPTION
**Pull request by zhexian**
has been tested

### CODE REVIEW

File: library/src/main/java/com/skocken/efficientadapter/lib/adapter/AdapterHelper.java [hunk 1]
```diff
@@ -293,31 +293,39 @@ public boolean onLongClick(View v) {
...
+                    return (EfficientViewHolder) viewHolder;
                 }
+
+                // inner class ViewHolder inside EfficientAdapter
+                if (parameterTypes.length == 2 && EfficientAdapter.class.isAssignableFrom(parameterTypes[0]) && View.class.isAssignableFrom(parameterTypes[1])) {
+                    Object viewHolder = constructor.newInstance(adapter, v);
+                    return (EfficientViewHolder) viewHolder;
+                }
+            } catch (Exception e) {
+                e.printStackTrace();
```
Comment (StanKocken) [inline library/src/main/java/com/skocken/efficientadapter/lib/adapter/AdapterHelper.java]: The "e.printStackTrace();" is not necessary as you throw a new Exception with it.

File: library/src/main/java/com/skocken/efficientadapter/lib/adapter/AdapterHelper.java [hunk 2]
```diff
@@ -293,31 +293,39 @@ public boolean onLongClick(View v) {
...
         Constructor<?>[] constructors = viewHolderClass.getDeclaredConstructors();
-        if (constructors != null) {
-            for (Constructor<?> constructor : constructors) {
-                Class<?>[] parameterTypes = constructor.getParameterTypes();
-                if (parameterTypes != null
-                        && parameterTypes.length == 1
-                        && parameterTypes[0].isAssignableFrom(View.class)) {
-                    return constructor;
+
+        if (constructors == null)
```
Comment (StanKocken) [inline library/src/main/java/com/skocken/efficientadapter/lib/adapter/AdapterHelper.java]: In term of coding style, I prefer the use of braces unless the condition and the body fit on one line.

File: library/src/main/java/com/skocken/efficientadapter/lib/adapter/AdapterHelper.java [hunk 3]
```diff
@@ -293,31 +293,39 @@ public boolean onLongClick(View v) {
...
+            throw new IllegalArgumentException(
+                    "Impossible to found a constructor with a view for "
+                            + viewHolderClass.getSimpleName());
+
+        for (Constructor<?> constructor : constructors) {
+            Class<?>[] parameterTypes = constructor.getParameterTypes();
+            if (parameterTypes == null)
```
Comment (StanKocken) [inline library/src/main/java/com/skocken/efficientadapter/lib/adapter/AdapterHelper.java]: Same comment about the braces here

File: library/src/main/java/com/skocken/efficientadapter/lib/adapter/AdapterHelper.java [hunk 4]
```diff
@@ -293,31 +293,39 @@ public boolean onLongClick(View v) {
...
+
+            try {
+                //single or static inner class ViewHolder
+                if (parameterTypes.length == 1 && View.class.isAssignableFrom(parameterTypes[0])) {
+                    Object viewHolder = constructor.newInstance(v);
+                    return (EfficientViewHolder) viewHolder;
                 }
+
+                // inner class ViewHolder inside EfficientAdapter
+                if (parameterTypes.length == 2 && EfficientAdapter.class.isAssignableFrom(parameterTypes[0]) && View.class.isAssignableFrom(parameterTypes[1])) {
```
Comment (StanKocken) [inline library/src/main/java/com/skocken/efficientadapter/lib/adapter/AdapterHelper.java]: Coding style: avoid line length over 100 chars.

**Comment by StanKocken** [comment]
Thanks for the PR !
Your code is correct and I will be happy to integrate it into the library.

I put some comment about coding style corrections to do.
I will also ask you to squash your commits together so that the repository history is clean.

Thanks again.

File: library/src/main/java/com/skocken/efficientadapter/lib/adapter/AdapterHelper.java (hunk 1 above)
Comment (zhexian) [inline library/src/main/java/com/skocken/efficientadapter/lib/adapter/AdapterHelper.java]: e.printStackTrace(); is trace for the real reason that case this exception,how do you think?

Comment (StanKocken) [inline library/src/main/java/com/skocken/efficientadapter/lib/adapter/AdapterHelper.java]: When you throw the RuntimeException, you provide the "reason", which is "e". So the stacktrace will display it anyway.

http://stackoverflow.com/questions/7469316/why-is-exception-printstacktrace-considered-bad-practice

Comment (zhexian) [inline library/src/main/java/com/skocken/efficientadapter/lib/adapter/AdapterHelper.java]: OK,get it

**Pull request closed without merging**